src/device_manager.rs

```diff
@@ -0,0 +1,359 @@
+use std::collections::btree_map::BTreeMap;
+use std::result;
+use std::sync::Arc;
+
+use vm_memory::GuestAddress;
+
+/// Error type for `IoManager` usage.
+#[derive(Debug)]
+pub enum Error {
+    /// The inserting device overlaps with a current device.
+    DeviceOverlap,
+    /// The device doesn't exist.
+    DeviceNotExist,
+}
+
+/// Simplify the `Result` type.
+pub type Result<T> = result::Result<T, Error>;
+
+// Structure describing an IO range.
+#[derive(Debug, Copy, Clone)]
+struct IoRange {
+    base: IoAddress,
+    size: IoSize,
+}
+
+impl IoRange {
+    fn new_pio_range(base: u16, size: u16) -> Self {
+        IoRange {
+            base: IoAddress::Pio(base),
+            size: IoSize::Pio(size),
+        }
+    }
+    fn new_mmio_range(base: u64, size: u64) -> Self {
+        IoRange {
+            base: IoAddress::Mmio(GuestAddress(base)),
+            size: IoSize::Mmio(size),
+        }
+    }
+}
+
+impl Eq for IoRange {}
+
+impl PartialEq for IoRange {
+    fn eq(&self, other: &IoRange) -> bool {
+        self.base == other.base
+    }
+}
+
+impl Ord for IoRange {
+    fn cmp(&self, other: &IoRange) -> Ordering {
+        self.base.cmp(&other.base)
+    }
+}
+
+impl PartialOrd for IoRange {
+    fn partial_cmp(&self, other: &IoRange) -> Option<Ordering> {
+        self.base.partial_cmp(&other.base)
+    }
+}
+
+/// System IO manager serving for all devices management and VM exit handling.
+#[derive(Default)]
+pub struct IoManager {
+    /// Range mapping for VM exit pio operations.
+    pio_bus: BTreeMap<IoRange, Arc<dyn DeviceIo>>,
+    /// Range mapping for VM exit mmio operations.
+    mmio_bus: BTreeMap<IoRange, Arc<dyn DeviceIo>>,
+}
+
+impl IoManager {
+    /// Create an default IoManager with empty IO member.
+    pub fn new() -> Self {
+        IoManager::default()
+    }
+    /// Register a new device IO with its allocated resources.
+    /// VMM is responsible for providing the allocated resources to virtual device.
+    ///
+    /// # Arguments
+    ///
+    /// * `device`: device instance object to be registered
+    /// * `resources`: resources that this device owns, might include
+    ///                port I/O and memory-mapped I/O ranges, irq number, etc.
+    pub fn register_device_io(
+        &mut self,
+        device: Arc<dyn DeviceIo>,
+        resources: &[Resource],
+    ) -> Result<()> {
+        // Register and mark device resources
+        // The resources addresses being registered are sucessfully allocated before.
+        for (idx, res) in resources.iter().enumerate() {
+            match *res {
+                Resource::PioAddressRange { base, size } => {
+                    if self
+                        .pio_bus
+                        .insert(IoRange::new_pio_range(base, size), device.clone())
+                        .is_some()
+                    {
+                        // Unregister registered resources.
+                        self.unregister_device_io(&resources[0..idx])?;
```

Comment:
Should we use expect() or unwrap() here?
The unregister_device_io() shouldn't fail here, otherwise we are already in troubles.